modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GridH2Table.java

```diff
@@ -1291,6 +1313,107 @@ public void removeIndex(Session session, Index h2Idx) {
+     *
+     * @throws IgniteCheckedException In case we were unable to destroy the data.
+     */
+    public void prepareIndexesForRebuild(Session session) throws IgniteCheckedException {
+        lock(true);
+
+        try {
+            ArrayList<Index> newIdxs = new ArrayList<>(idxs.size());
+
+            for (int i = 0; i < sysIdxsCnt; i++) {
+                Index idx = idxs.get(i);
+
+                if (idx instanceof GridH2ProxyIndex)
+                    break;
+
+                Index newIdx = idx instanceof H2TreeIndex ? recreateIndex((H2TreeIndex) idx) : idx;
+
+                newIdxs.add(newIdx);
+            }
+
+            for (int i = 1; i < sysIdxsCnt; i++) {
+                Index clone = createDuplicateIndexIfNeeded(newIdxs.get(i));
+
+                if (clone != null)
+                    newIdxs.add(clone);
+            }
+
+            List<IgniteBiTuple<Index, Index>> toReplace = new ArrayList<>();
+
+            for (int i = sysIdxsCnt; i < idxs.size(); i++) {
+                Index idx = idxs.get(i);
+
+                if (idx instanceof GridH2ProxyIndex)
+                    continue;
+
+                if (idx instanceof H2TreeIndex) {
+                    Index newIdx = recreateIndex((H2TreeIndex) idx);
+
+                    newIdxs.add(newIdx);
+
+                    toReplace.add(new IgniteBiTuple<>(idx, newIdx));
+
+                    Index clone = createDuplicateIndexIfNeeded(newIdx);
+
+                    if (clone != null) {
+                        newIdxs.add(clone);
+
+                        toReplace.add(new IgniteBiTuple<>(null, clone));
+                    }
+                }
+                else {
+                    newIdxs.add(idx);
+
+                    if (idxs.get(i + 1) instanceof GridH2ProxyIndex)
+                        newIdxs.add(idxs.get(++i));
+                }
+            }
+
+            for (IgniteBiTuple<Index, Index> oldToNew : toReplace)
+                replaceSchemaObject(session, oldToNew.get1(), oldToNew.get2());
+
+            if (hasHashIndex)
+                newIdxs.set(0, new H2TableScanIndex(this, (GridH2IndexBase) newIdxs.get(2), (GridH2IndexBase) newIdxs.get(1)));
+            else
+                newIdxs.set(0, new H2TableScanIndex(this, (GridH2IndexBase) newIdxs.get(1), null));
+
+            idxs = newIdxs;
+
+            incrementModificationCounter();
+        }
+        finally {
+            unlock(true);
+        }
+    }
+
+    /** */
+    private H2TreeIndex recreateIndex(H2TreeIndex treeIdx) throws IgniteCheckedException {
+        treeIdx.destroy0(true, true);
+
+        GridCacheContext<?, ?> cctx = cacheContext();
+
+        assert cctx != null;
+
+        return treeIdx.createCopy(cctx.dataRegion().pageMemory(), cctx.offheap());
+    }
+
+    /** */
+    private void replaceSchemaObject(
+            Session session,
+            @Nullable SchemaObject oldObj,
+            @Nullable SchemaObject newObj
+    ) {
+        if (oldObj != null)
```

Comment:
Let's add an assert that lock is held by current thread.

Reply:
good idea!